server/src/routes/groupExpenses.ts

```diff
@@ -0,0 +1,213 @@
+      (sum, share) => sum + share,
+      0,
+    );
+    const EPSILON = 0.0001;
+    if (Math.abs(totalShare - 1) > EPSILON) {
+      throw new Error(
+        'Split shares must add up to 1 within a small margin of error',
+      );
+    }
+    return true;
+  }),
+  async (req: Request, res: Response) => {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      return res.status(400).json({ errors: errors.array() });
+    }
+
+    try {
+      const { title, amount, createdAt, category, split } = req.body;
+      const usernames = Object.keys(split);
+
+      // Verify that the user that made the request exists
+      const user = await prisma.user.findUnique({
+        where: { id: req.userId },
+      });
+
+      if (!user || user.userDeleted) {
+        return res.status(401).json({ message: 'Unauthorized' });
+      }
+
+      // Verify users exist
+      const users = await prisma.user.findMany({
+        where: { username: { in: usernames } },
+      });
+
+      if (users.length !== usernames.length) {
+        return res.status(400).json({
+          message: "Invalid usernames, at least one of the users doesn't exist",
+        });
+      }
+
+      // Verify that none of the users were deleted
+      const deletedUsers = users.filter((user) => user.userDeleted);
+      if (deletedUsers.length > 0) {
+        return res.status(400).json({
+          message: "Invalid usernames, at least one of the users doesn't exist",
+        });
+      }
+
+      // Create the group expense
+      const groupExpense = await prisma.groupExpense.create({
+        data: {
+          title,
+          totalAmount: amount,
+          createdAt,
+          category,
+        },
+      });
+
+      // Create the group expense splits
+      for (const user of users) {
+        await prisma.groupExpenseSplit.create({
+          data: {
+            userId: user.id,
+            groupExpenseId: groupExpense.id,
+            shareAmount: split[user.username] * amount,
+            hasPaid: false,
+          },
+        });
+      }
+
+      res.status(201).json({
+        message: 'Group expense successfully created',
+        groupExpense,
+      });
+    } catch (error) {
+      res.locals.error = error;
+      res.status(500).json({ message: 'Server error' });
+    }
+  },
+);
+
+/**
+ * Mark a group expense as paid
+ *
+ * This route marks a group expense as paid for the authenticated user.
+ * The user must be part of the group expense.
+ * Once the group expense is marked as paid, a new expense is created for the user.
+ *
+ * @route PUT /group-expenses/:groupExpenseId/mark-as-paid
+ * @param groupExpenseId ID of the group expense
+ * @returns {object} An object containing the newly created group expense
+ * @throws {object} 401 - If the user is not authenticated
+ * @throws {object} 404 - If the group expense is not found
+ * @throws {object} 500 - If there is a server error
+ */
```

Comment:
Just like a question, this marks the whole group as paid, I thought it was for each user marked as paid, but I totally understand if its too much work

Reply:
The comment here is inaccurate, I'll change it. This endpoint only marks the group expense as paid for the user that made the request.